lib/slimmer/processors/feedback_url_swapper.rb

```diff
@@ -0,0 +1,32 @@
+module Slimmer::Processors
+  class FeedbackURLSwapper
+    def initialize(request, headers = {})
+      @headers = headers
+      @request = request
+    end
+
+    def filter(_src, dest)
+      email_regex = /[^\s=\/?&]+(?:@|%40)[^\s=\/?&]+/
+
+      original_url = @request.base_url + @request.fullpath
+      original_url_without_pii = utf_encode(original_url.gsub(email_regex, "[email]"))
+      dest.at_css(".gem-c-feedback input[name='url']").set_attribute("value", original_url_without_pii) if is_gem_layout?
+
+      full_path = @request.fullpath
+      full_path_without_pii = utf_encode(full_path.gsub(email_regex, "[email]"))
+      dest.at_css(".gem-c-feedback input[name='email_survey_signup[survey_source]']").set_attribute("value", full_path_without_pii) if is_gem_layout?
+
+      dest
+    end
+
+  private
+
+    def utf_encode(element)
+      element.is_a?(String) ? element.encode : element
```

Comment:
I've been doing some experimenting and I don't think this encode method is needed. 

I found out a couple of things:

1. Just using `.encode` isn't ideal as it doesn't fix any characters it doesn't recognise and [that it's behaviour could be surprising in some scenarios](https://makandracards.com/makandra/41250-ruby-s-default-encodings-can-be-unexpected) - so it should use `force_encoding("UTF-8")` rather than just `.encode
2. As Nokogiri seems to sort it out anyway (unlike ERB in the templates) it seems impossible to set-up a test that will fail if we don't do the encoding.

I'd suggest rather than having any code for it we perhaps just have a test that it can cope with different coded string and skip the logic. This was the test I was using:

```ruby
  def test_should_cope_with_a_ascii_encoded_url
    template = as_nokogiri %(
      <div class="gem-c-feedback">
        <input id="test-input" name="email_survey_signup[survey_source]" value="/old_path">
        <input id="test-input-two" name="url" value="https://example.com/old_path">
      </div>
    )

    # need to use a stub, as a MockRequest will exception on ASCII characters
    request = stub("Rack::Request",
                   base_url: "https://example.com".force_encoding("ASCII-8BIT"),
                   fullpath: "/test?áscii=%EE%90%80".force_encoding("ASCII-8BIT"))
    headers = { Slimmer::Headers::TEMPLATE_HEADER => "gem_layout" }

    Slimmer::Processors::FeedbackURLSwapper.new(request, headers).filter(nil, template)

    assert_in template, "#test-input[value='/test?áscii=%EE%90%80']"
    assert_in template, "#test-input-two[value='https://example.com/test?áscii=%EE%90%80']"
  end
```

Reply:
I _think_ I followed what you've suggested - I've removed the check if it's a string, used `force_encoding`, and added the test to make sure that it all works. Is that right?

Reply:
Yup, that sounds good and sufficient 👍